src/soundsourceproxy.cpp

```diff
@@ -433,9 +433,69 @@ QImage SoundSourceProxy::parseCoverImage() const {
     return coverImg;
 }
 
+namespace {
+
+// Keeps the TIO alive while accessing the audio data
+// of the track. The TIO must not be deleted while
+// accessing the corresponding file to avoid file
+// corruption when writing metadata while the file
+// is still in use.
+class AudioSourceProxy: public Mixxx::AudioSource {
+public:
+    AudioSourceProxy(const AudioSourceProxy&) = delete;
+    AudioSourceProxy(AudioSourceProxy&&) = delete;
+
+    static Mixxx::AudioSourcePointer create(
+            const TrackPointer& pTrack,
+            const Mixxx::AudioSourcePointer& pAudioSource) {
+        DEBUG_ASSERT(!pTrack.isNull());
+        DEBUG_ASSERT(!pAudioSource.isNull());
+        return Mixxx::AudioSourcePointer(
+                new AudioSourceProxy(pTrack, pAudioSource));
+    }
+
+    AudioSourceProxy(
+            const TrackPointer& pTrack,
+            const Mixxx::AudioSourcePointer& pAudioSource)
+        : Mixxx::AudioSource(*pAudioSource),
+          m_pTrack(std::move(pTrack)),
```

Comment:
I do not understand why you use move here here. Maybe I have not yet understood the move semantic an rvalue reference thing. So please correct me if I am wrong. 

std::move casts the const TrackPointer& pTrack which can be either an lvalue reference or to rvalue reference?  Expressed in C++03 is may look like 
`TrackPointer move(const TrackPointer& pTrack) {return pTrack;}` 
Since pTrack is const, the compiler has to create a new rvalue in the air in any case, even if pTrack was originally already a reference to an rvalue, which resists in the air outside of this function. 
```
const Track& outside = getTrack(); // <- outside is an rvalue reference 
new AudioSourceProxy(outside) ; 
```
If you change the interface to TrackPointer&& pTrack it would make more sense. Then we would actually move the original rvalue of getTrack() to the new m_pTrack using std::move(pTrack)
```
Track&& outside = getTrack(); // <- outside is an rvalue reference 
new AudioSourceProxy(outside) ; 
```

The other issue: 
QSharedPointer has no move constructor and since it has a user defined copy constructor there will also no compiler generated one. So in our case, std::move generates a new rvalue instead of just passing the unchanged parameter 
to QSharedPointer(const QSharedPointer<T> & other)